design/custom-request-timeout.md

```diff
@@ -0,0 +1,256 @@
+	PerTryTimeout string `json:"perTryTimeout"`
+}
+```
+
+## Intermediate form in the DAG
+We store unmarshalled structs from ingress' and ingressroute's yaml into a common dag struct route for both. 
+There are three approaches to do change:
+
+### Approach 1
+Just add IdleTimeout as another member in the struct as below as Idle Timeout is the only new member
+in the _current_ proposed CRD.
+
+```golang
+// Approach 1: contour/internal/dag/dag.go
+
+type Route struct {
+    [... other members ...]
+    IdleTimeout time.Duration
+}
+
+```
+
+### Approach 2
+
+Consolidate timeout and retry members of dag's Route struct to their own respective struct.
+
+```golang
+// Approach 2: contour/internal/dag/dag.go
+
+type RouteTimeout struct {
+    Request time.Duration
+    Idle    time.Duration
+    [... more members can be added based on requirements ...]
+}
+
+type RouteRetry struct {
+    MaxRetries      int
+    OnStatusCodes   []string
+    PerTryTimeout   time.Duration
+    [... more members can be added based on requirements ...]
+}
+
+type Route struct {
+    Prefix          string
+    object          interface{}
+    httpServices    map[servicemeta]*HTTPService
+    HTTPSUpgrade    bool
+    Websocket       bool
+    PrefixRewrite   string
+
+    // new member replacing current Timeout
+    Timeout         RouteTimeout
+    // new member replacing current RetryOn, NumRetries, PerTryTimeout
+    Retry           RouteRetry
+}
+```
+
+This could involve more changes in the code and their unit tests to be corrected, but hopefully,
+the future additions if made can be neatly encapsulated.
+
+### Approach 3
+No change in the dag's Route struct, pick the required new member from the `object` interface when needed.
+
+Depending on the above changes, `prefixRoute` function for Ingress object and `processRoutes` function
+for IngressRoute object would need to be visited for filling in the decided Route structure.
+
+## Communicating from DAG to Envoy
+
+Now that we have the DAG built, we need to communicate the state of the routes to Envoy. This happens `RouteRoute`
+function of `contour/internal/envoy/route.go` 
+
+We would use the following to map value from DAG's route to protobuf
+
+```
+RouteAction.Timeout = Route.Timeout.Request
+RouteAction.IdleTimeout = Route.Timeout.Idle
+
+RouteAction.RetryPolicy.RetryOn = Route.Retry.OnStatusCodes // list joined into a string
+RouteAction.RetryPolicy.NumEntries = Route.Retry.MaxRetries
+RouteAction.RetryPolicy.PerTryTimeout = Route.Retry.PerTryTimeout
+
+```
+
+## Validation rules
+
+Let's consider the following topology for communication
+
++---------+      +---------+       +----------+
+
+| Client  |----->| Proxy   |------>| Server   |
+
++---------+      +---------+       +----------+
+
+
+### Route.Timeout
+
+- `Request` : The time that spans between the point at which complete client request has been processed by the proxy, and when the
```

Comment:
Could you link Envoy docs for the field definition? Less moving parts that way. Thanks!

Reply:
Will do that.